crates/api/src/id.rs

```diff
@@ -0,0 +1,288 @@
+//! Types dealing with data identity or hashing.
+
+macro_rules! imp_deref {
+    ($i:ty, $t:ty) => {
+        impl std::ops::Deref for $i {
+            type Target = $t;
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+    };
+}
+
+macro_rules! imp_from {
+    ($a:ty, $b:ty, $i:ident => $e:expr) => {
+        impl From<$b> for $a {
+            fn from($i: $b) -> Self {
+                $e
+            }
+        }
+    };
+}
+
+/// Base data identity type meant for newtyping.
+/// You probably want [AgentId] or [OpId].
+///
+/// In Kitsune2 these bytes should ONLY be the actual hash bytes
+/// or public key of the identity being tracked, without
+/// prefix or suffix.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Id(pub bytes::Bytes);
+
+impl serde::Serialize for Id {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use base64::prelude::*;
+        serializer.serialize_str(&BASE64_URL_SAFE_NO_PAD.encode(&self.0))
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for Id {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use base64::prelude::*;
+        let s: &'de str = serde::Deserialize::deserialize(deserializer)?;
+        BASE64_URL_SAFE_NO_PAD
+            .decode(s)
+            .map(|v| Id(bytes::Bytes::copy_from_slice(&v)))
+            .map_err(serde::de::Error::custom)
+    }
+}
+
+imp_deref!(Id, bytes::Bytes);
+imp_from!(Id, bytes::Bytes, b => Id(b));
+
+impl Id {
+    /// Get the location u32 based off this Id.
+    ///
+    /// This is accomplished by directly xor-ing every successive 4 bytes
+    /// in the hash. It is okay if the hash len is not a multiple of 4,
+    /// it will stop with the ending byte of the hash.
+    ///
+    /// The remaining 4 bytes are then interpreted as a little-endian u32.
+    //
+    // Holochain previously would re-hash the hash, and then
+    // xor to shrink down to a u32. This extra step is not needed
+    // and does not provide any benefit. One extra hash step does
+    // not prevent location farming, and if the original hash was
+    // distributed well enough, re-hashing it again doesn't improve
+    // distribution.
+    pub fn loc(&self) -> u32 {
+        let mut out = [0_u8; 4];
+        let mut i = 0;
+        for c in &self.0 {
+            out[i] ^= c;
```

Comment:
I would find this clearer as a ranged loop with indexing on the modulo of the index. 

```
for i in 0..self.0.len() {
  out[i % 4] ^= self.0[i]
}
```

Or something along those lines, I haven't checked that code runs